File: modules/go/nyx/entities/verbosity.go
/*
 * Copyright 2020 Mooltiverse
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package entities

import (
	"fmt" // https://pkg.go.dev/fmt

	log "github.com/sirupsen/logrus" // https://pkg.go.dev/github.com/sirupsen/logrus

	errs "github.com/mooltiverse/nyx/modules/go/errors"
)

/*
This type maps log levels and the corresponding configuration options and state attributes,
whose string representations may not always match the log level values.

In other words, here is the mapping between logging levels and the string values users can use in configuration
and state to ask for those logging levels.
*/
type Verbosity string

const (
	// The fatal log level.
	FATAL Verbosity = "FATAL"

	// The error log level.
	ERROR Verbosity = "ERROR"

	// The warning log level.
	WARNING Verbosity = "WARNING"

	// The info log level.
	INFO Verbosity = "INFO"

	// The debug log level.
	DEBUG Verbosity = "DEBUG"

	// The trace log level.
	TRACE Verbosity = "TRACE"
)

/*
Returns the logging level corresponding to this verbosity level
*/
func (v Verbosity) GetLevel() log.Level {
	switch v {
	case FATAL:
		return log.FatalLevel
	case ERROR:
		return log.ErrorLevel
	case WARNING:
		return log.WarnLevel
	case INFO:
		return log.InfoLevel
	case DEBUG:
		return log.DebugLevel
	case TRACE:
		return log.TraceLevel
	default:
		// this is never reached, but in case, return INFO as the default
		return log.InfoLevel
	}
}

/*
Returns the verbosity corresponding to the given logging level.

Errors can be:

- IllegalPropertyError in case an unknown log level is passed
*/
func FromLevel(l log.Level) (Verbosity, error) {
	switch l {
	case log.PanicLevel, log.FatalLevel:
		return FATAL, nil
	case log.ErrorLevel:
		return ERROR, nil
	case log.WarnLevel:
		return WARNING, nil
	case log.InfoLevel:
		return INFO, nil
	case log.DebugLevel:
		return DEBUG, nil
	case log.TraceLevel:
		return TRACE, nil
	default:
		// this is never reached, but in case, return INFO as the default, with an error
		return INFO, &errs.IllegalPropertyError{Message: fmt.Sprintf("illegal log level '%s'", l)}
	}
}

/*
Returns the string representation of the verbosity level
*/
func (v Verbosity) String() string {
	switch v {
	case FATAL:
		return "FATAL"
	case ERROR:
		return "ERROR"
	case WARNING:
		return "WARNING"
	case INFO:
		return "INFO"
	case DEBUG:
		return "DEBUG"
	case TRACE:
		return "TRACE"
	default:
		// this is never reached, but in case...
		panic("unknown Verbosity. This means the switch/case statement needs to be updated")
	}
}

/*
Returns the verbosity corresponding to the given string.

Errors can be:

- IllegalPropertyError in case an unknown log level is passed
*/
func ValueOfVerbosity(s string) (Verbosity, error) {
	switch s {
	case "FATAL":
		return FATAL, nil
	case "ERROR":
		return ERROR, nil
	case "WARNING":
		return WARNING, nil
	case "INFO":
		return INFO, nil
	case "DEBUG":
		return DEBUG, nil
	case "TRACE":
		return TRACE, nil
	default:
		return INFO, &errs.IllegalPropertyError{Message: fmt.Sprintf("illegal log level '%s'", s)}
	}
}

File: src/java/main/src/main/java/com/mooltiverse/oss/nyx/entities/Verbosity.java
/*
 * Copyright 2020 Mooltiverse
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.mooltiverse.oss.nyx.entities;

import java.util.Objects;

import org.slf4j.event.Level;

/**
 * This class maps log {@link Level}s and the corresponding configuration options and state attributes,
 * whose string representations may not always match the {@link Level} values.
 * 
 * In other words, here is the mapping between logging levels and the string values users can use in configuration
 * and state to ask for those logging levels.
 */
public enum Verbosity {
    /**
     * The fatal log level. Please note that this level has no corresponding level into SLF4J
     * so the value is the same as {@link #ERROR}.
     * 
     * @see Level#ERROR
     */
    FATAL(Level.ERROR),

    /**
     * The error log level.
     * 
     * @see Level#ERROR
     */
    ERROR(Level.ERROR),

    /**
     * The warning log level.
     * 
     * @see Level#WARN
     */
    WARNING(Level.WARN),

    /**
     * The info log level.
     * 
     * @see Level#INFO
     */
    INFO(Level.INFO),

    /**
     * The debug log level.
     * 
     * @see Level#DEBUG
     */
    DEBUG(Level.DEBUG),

    /**
     * The trace log level.
     * 
     * @see Level#TRACE
     */
    TRACE(Level.TRACE);

    /**
     * The SLF4J corresponding logging level.
     */
    private final Level level;

    /**
     * Builds the enumeration item.
     * 
     * @param level the SLF4J corresponding logging level
     */
    private Verbosity(Level level) {
        this.level = level;
    }

    /**
     * Returns the SLF4J level corresponding to this verbosity level
     * 
     * @return the SLF4J level corresponding to this verbosity level
     */
    public Level getLevel() {
        return level;
    }

    /**
     * Returns the proper verbosity level mapped from the given SLF4J level.
     * 
     * @param level the SLF4J level to parse and return the verbosity for
     * 
     * @return the proper verbosity level mapped from the given SLF4J level.
     * 
     * @throws IllegalArgumentException if the given value cannot be mapped to any existing verbosity level
     * @throws NullPointerException if the given value is {@code null}
     */
    public static Verbosity fromLevel(Level level)
        throws IllegalArgumentException, NullPointerException {
        if (Objects.isNull(level))
            throw new NullPointerException();

        switch (level) {
            case TRACE: return Verbosity.TRACE;
            case DEBUG: return Verbosity.DEBUG;
            case INFO:  return Verbosity.INFO;
            case WARN:  return Verbosity.WARNING;
            case ERROR: return Verbosity.ERROR;
            default: throw new IllegalArgumentException(level.toString());
        }
    }
}
